kedro/io/kedro_data_catalog.py

```diff
@@ -231,30 +273,31 @@ def _get_dataset(
     def add(
         self, ds_name: str, dataset: AbstractDataset, replace: bool = False
     ) -> None:
+        # TODO: remove when removing old catalog
         """Adds a new ``AbstractDataset`` object to the ``KedroDataCatalog``."""
-        if ds_name in self._datasets:
-            if replace:
-                self._logger.warning("Replacing dataset '%s'", ds_name)
-            else:
-                raise DatasetAlreadyExistsError(
-                    f"Dataset '{ds_name}' has already been registered"
-                )
-        self._datasets[ds_name] = dataset
-
-    def list(self, regex_search: str | None = None) -> list[str]:
+        if ds_name in self._datasets and not replace:
+            raise DatasetAlreadyExistsError(
+                f"Dataset '{ds_name}' has already been registered"
+            )
+        self.__setitem__(ds_name, dataset)
+
+    def list(self, regex_search: str | None = None) -> List[str]:  # noqa: UP006
+        # TODO: remove when removing old catalog
         """
         List of all dataset names registered in the catalog.
         This can be filtered by providing an optional regular expression
         which will only return matching keys.
         """
+        if regex_search == "":
+            self._logger.warning("The empty string will not match any datasets")
+            return []
+
+        return self.keys(regex_search)
 
+    def _filter_keys(self, regex_search: str | None) -> List[str]:  # noqa: UP006
         if regex_search is None:
             return list(self._datasets.keys())
 
-        if not regex_search.strip():
-            self._logger.warning("The empty string will not match any datasets")
-            return []
-
         try:
             pattern = re.compile(regex_search, flags=re.IGNORECASE)
```

Comment:
if ``regex_search`` is forced to be a string, should we add ``regex_kwargs`` everywhere to handle these flags?